# ¡Feliz Navidad a toda LATAM! 🎄✨

Soy **Hesus Garcia** – tu revisor de confianza y ahora también el **Hesus que trae la magia navideña al código** 🎅🦸‍♂️. Sí, soy el único Hesus con H que viene cargado de observaciones, sugerencias y buenos deseos para ayudarte a mejorar tu proyecto en estas fechas tan especiales. 🎁

Quiero aprovechar esta oportunidad para desearte a ti y a toda nuestra increíble comunidad de estudiantes en Latinoamérica unas fiestas llenas de alegría, aprendizaje y nuevos logros. 🌟 ¡Que esta Navidad esté llena de código limpio, proyectos exitosos y muchos "outputs" perfectos! 🚀

Cada vez que encuentre un detalle importante en tu código, te lo señalaré con la intención de que puedas corregirlo y brilles en tu camino profesional. Recuerda que estos comentarios no son más que regalos disfrazados, pensados para ayudarte a crecer y mejorar. 🎁 Si alguna vez te encuentras con un problema que no logras resolver, no te preocupes; estaré aquí para guiarte.

Es importante que cuando encuentres un comentario, **no los muevas, no los modifiques, ni los borres**. Estos son como los adornos de un árbol de Navidad 🎄; cada uno está puesto en su lugar por una razón. 

---

### **Formato de Comentarios**

Revisaré cuidadosamente cada implementación en tu notebook para asegurar que cumpla con los requisitos y te daré comentarios de acuerdo al siguiente formato:

---

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
    
<b>Éxito</b> - ¡Excelente trabajo! Esta parte está bien implementada y contribuye significativamente al análisis de datos o al proyecto. Continúa aplicando estas buenas prácticas en futuras secciones. 🎉
    
</div>

<div class="alert alert-block alert-warning">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
    
<b>Atención</b> ⚠️ - Este código está correcto, pero se puede optimizar. Considera implementar mejoras para que sea más eficiente y fácil de leer. Esto fortalecerá la calidad de tu proyecto. 🔧
    
</div>

<div class="alert alert-block alert-danger">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
    
<b>A resolver</b> ❗ - Aquí hay un problema o error en el código que es necesario corregir para aprobar esta sección. Por favor, revisa y corrige este punto, ya que es fundamental para la validez del análisis y la precisión de los resultados. 🚨
    
</div>

---

Al final de cada revisión, recibirás un **Comentario General del Revisor**, como un resumen festivo de los aspectos positivos y áreas para mejorar en tu proyecto. Esto incluirá:

- **Aspectos positivos:** Un resumen de los puntos fuertes de tu proyecto. 🎯
- **Áreas de mejora:** Sugerencias donde puedes mejorar. 🔍
- **Temas adicionales para investigar:** Ideas opcionales para expandir tus habilidades. 🌱

---

Espero que este espíritu navideño te motive a seguir adelante y que este proyecto sea uno de muchos logros en tu camino. ¡Felicidades por tu dedicación y esfuerzo! 💪🎅✨

### ¡Empecemos con esta revisión navideña! 🎄🚀

# Megaline. Machine Learning

# Descripción del proyecto

La compañía móvil Megaline no está satisfecha al ver que muchos de sus clientes
utilizan planes heredados. Quieren desarrollar un modelo que pueda analizar el
comportamiento de los clientes y recomendar uno de los nuevos planes de Megaline:
Smart o Ultra.
Para esta tarea de clasificación debes crear un modelo que escoja el plan correcto.
Para ayudarte a entrenar tu modelo, utilizarás los datos de comportamiento del usuario
del curso sobre Análisis estadísticos de datos. Como ya hiciste el paso de procesar los
datos, puedes lanzarte directo a crear el modelo.
Desarrolla un modelo con la mayor exactitud posible. En este proyecto, el umbral de
exactitud es 0.75. Usa el dataset para comprobar la exactitud.

In [44]:
#importar librerias

import pandas as pd
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import mean_squared_error 
from sklearn.ensemble import RandomForestRegressor
from sklearn.dummy import DummyClassifier

In [45]:
#importar librerias 
df = pd.read_csv('/datasets/users_behavior_upd.csv')

print(df.shape)

(3214, 5)


In [46]:
df

calls  minutes  messages   mb_used  is_ultimate
0      40.0   311.90      83.0  19915.42            0
1      85.0   516.75      56.0  22696.96            0
2      77.0   467.66      86.0  21060.45            0
3     106.0   745.53      81.0   8437.39            1
4      66.0   418.74       1.0  14502.75            0
...     ...      ...       ...       ...          ...
3209  122.0   910.98      20.0  35124.90            1
3210   25.0   190.36       0.0   3275.61            0
3211   97.0   634.44      70.0  13974.06            0
3212   64.0   462.32      90.0  31239.78            0
3213   80.0   566.09       6.0  29480.52            1

[3214 rows x 5 columns]

In [47]:
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   calls        3214 non-null   float64
 1   minutes      3214 non-null   float64
 2   messages     3214 non-null   float64
 3   mb_used      3214 non-null   float64
 4   is_ultimate  3214 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 125.7 KB


In [48]:
df.describe()

calls      minutes     messages       mb_used  is_ultimate
count  3214.000000  3214.000000  3214.000000   3214.000000  3214.000000
mean     63.038892   438.208787    38.281269  17207.673836     0.306472
std      33.236368   234.569872    36.148326   7570.968246     0.461100
min       0.000000     0.000000     0.000000      0.000000     0.000000
25%      40.000000   274.575000     9.000000  12491.902500     0.000000
50%      62.000000   430.600000    30.000000  16943.235000     0.000000
75%      82.000000   571.927500    57.000000  21424.700000     1.000000
max     244.000000  1632.060000   224.000000  49745.730000     1.000000

In [49]:
df.duplicated().count()

3214

In [50]:
df.drop_duplicates()

calls  minutes  messages   mb_used  is_ultimate
0      40.0   311.90      83.0  19915.42            0
1      85.0   516.75      56.0  22696.96            0
2      77.0   467.66      86.0  21060.45            0
3     106.0   745.53      81.0   8437.39            1
4      66.0   418.74       1.0  14502.75            0
...     ...      ...       ...       ...          ...
3209  122.0   910.98      20.0  35124.90            1
3210   25.0   190.36       0.0   3275.61            0
3211   97.0   634.44      70.0  13974.06            0
3212   64.0   462.32      90.0  31239.78            0
3213   80.0   566.09       6.0  29480.52            1

[3214 rows x 5 columns]

# Conclusiones de los datos 

EL dataset que estaremos analizando es de una compania de megafonia. No tiene vaklores nulos y algunos de los datos son continuos, como la duración de las llamadas, mensajes, mb utilzados. El unico valor entero que es el plan

# Segmentación de datos

Segmenta los datos fuente en un conjunto de entrenamiento, uno de validación y
uno de prueba.

In [51]:
df_train, df_valid = train_test_split(df, test_size=0.20, random_state=12345)

In [52]:
features_train = df_train.drop(['is_ultimate'], axis=1)
target_train = df_train['is_ultimate']
features_valid = df_valid.drop(['is_ultimate'], axis=1)
target_valid = df_valid['is_ultimate']

<div class="alert alert-block alert-success">

<b>Comentario del revisor</b> <a class="tocSkip"></a><br>

<b>Éxito</b> - La segmentación de datos en conjuntos de entrenamiento y validación se realizó correctamente. Esto garantiza una evaluación confiable del desempeño de los modelos.
</div>

# Modelos

# Árbol de decisiones

In [53]:
for depth in range(1, 9): 
    model = DecisionTreeClassifier(random_state=12345, max_depth=depth)
    model.fit(features_train, target_train)

    predictions_valid = model.predict(features_valid)# < crea un bucle para max_depth de 1 a 5 >
    print("max_depth =", depth, ": ", end='')
    print(accuracy_score(target_valid, predictions_valid))

max_depth = 1 : 0.7480559875583204
max_depth = 2 : 0.7807153965785381
max_depth = 3 : 0.7838258164852255
max_depth = 4 : 0.7791601866251944
max_depth = 5 : 0.7853810264385692
max_depth = 6 : 0.7822706065318819
max_depth = 7 : 0.7791601866251944
max_depth = 8 : 0.7729393468118196


Conclusión.
Podemos ver en los resultados que un árbol con uns profundidad de 5 es la mejor para entrenar el  modelo , ya que si son más o menos su efectividad baja ya se por subajuste o sobreajuste del modelo

# Bosque Aleatorio

In [54]:
best_score = 0
best_est = 0
for est in range(1, 11): # selecciona el rango del hiperparámetro
    model = RandomForestClassifier(random_state=54321, n_estimators= est) # configura el número de árboles
    model.fit(features_train, target_train) # entrena el modelo en el conjunto de entrenamiento
    score = model.score(features_valid, target_valid) # calcula la puntuación de accuracy en el conjunto de validación
    if score > best_score:
        best_score = score# guarda la mejor puntuación de accuracy en el conjunto de validación
        best_est = est# guarda el número de estimadores que corresponden a la mejor puntuación de exactitud

print("La exactitud del mejor modelo en el conjunto de validación (n_estimators = {}): {}".format(best_est, best_score))

La exactitud del mejor modelo en el conjunto de validación (n_estimators = 10): 0.7807153965785381


<div class="alert alert-block alert-warning">

<b>Comentario del revisor</b> <a class="tocSkip"></a><br>

<b>Atención</b> ⚠️ - Gran trabajo optimizando n_estimators. Sin embargo, considera aumentar el rango del hiperparámetro para explorar configuraciones adicionales y mejorar potencialmente el desempeño del modelo.
</div>



# Regresión logística

In [55]:
model = LogisticRegression(random_state=54321, solver='liblinear')# inicializa el constructor de regresión logística con los parámetros random_state=54321 y solver='liblinear'
model.fit(features_train, target_train) # entrena el modelo en el conjunto de entrenamiento
score_train = model.score(features_train, target_train) # calcula la puntuación de accuracy en el conjunto de entrenamiento
score_valid = model.score(features_valid, target_valid) # calcula la puntuación de accuracy en el conjunto de validación

print("Exactitud del modelo de regresión logística en el conjunto de entrenamiento:", score_train)
print("Exactitud del modelo de regresión logística en el conjunto de validación:", score_valid)

Exactitud del modelo de regresión logística en el conjunto de entrenamiento: 0.7016725009723843
Exactitud del modelo de regresión logística en el conjunto de validación: 0.702954898911353


# Modelo de cordura

In [56]:
model = DummyClassifier(random_state = 42)
model.fit(features_train, target_train)
predictions = model.predict(features_train)
score_test = model.score(features_train, target_train)

print("Exactitud del modelo en el conjunto de prueba:", round(score_test, 5))

Exactitud del modelo en el conjunto de prueba: 0.69312


<div class="alert alert-block alert-success">

<b>Comentario del revisor</b> <a class="tocSkip"></a><br>

<b>Éxito</b> - El uso del DummyClassifier como punto de referencia es una buena práctica que demuestra profesionalismo en la evaluación de modelos.
</div>

# Conclusión de modelos de hiperparámetros

Árbol de decisiones = 0.7853810264385692
Bosque Aleatorio = 0.7807153965785381
Regresión logística = 0.7016725009723843
Modelo de cordura = 0.69312

con base en estos resultados, se hará la prueba de validación con el árbol de decisiones y bosque aleatorio ya que tuvieron los mejores valores de exactitud

In [59]:
#prueba de exactitud con bosque aleatorio 

final_model = RandomForestRegressor(random_state=54321, n_estimators=40)# inicializa el constructor para el modelo que tenía el mejor valor de RECM
final_model.fit(features_train, target_train) # entrena el modelo en el conjunto de entrenamiento
predictions = final_model.predict(features_valid)
result = mean_squared_error(target_valid, predictions)**0.5
print(result)

0.41012766854981736


<div class="alert alert-block alert-danger">

<b>Comentario del revisor</b> <a class="tocSkip"></a><br>

<b>A resolver</b> ❗ - En las pruebas finales, se utilizó RandomForestRegressor en lugar de RandomForestClassifier. Dado que el problema es de clasificación, esto podría haber afectado la evaluación. Reemplazarlo por RandomForestClassifier asegurará resultados más consistentes.
</div>

In [66]:
#prueba de exactitud con árbol de decisiones

final_model = None
final_model = 0

for depth in range(1, 6):
    model = DecisionTreeClassifier(random_state = 12345, max_depth = depth)
    model.fit(features_train, target_train)
    predictions = model.predict(features_valid)
    result = mean_squared_error(target_valid, predictions)**0.5
    
    if result > final_model:
        best_model = model
        final_model = result

print("Exactitud del mejor modelo en el conjunto de validación:", final_model )

Exactitud del mejor modelo en el conjunto de validación: 0.5019402478798444


# conclusión

se hizo la prueba de validación con el modelo de árbol de decisiones y con bosque aleatorio. Sin embargo el resultado de bosque aleatorio fue más bajo (0.41) por lo tanto es el modelo con el que nos quedaremos para validar el modelo. 

<div class="alert alert-block alert-warning">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

<b>Atención</b> ⚠️ - Para mejorar la claridad y profundidad de tus conclusiones finales, considera las siguientes pautas:

<ul>
    <li><b>Resumen de los Hallazgos Clave:</b> Incluye brevemente tus descubrimientos más importantes.</li>
    <li><b>Impacto de tus Elecciones de Procesamiento de Datos:</b> Reflexiona sobre cómo tus decisiones influyeron en los resultados.</li>
    <li><b>Relación con los Objetivos Iniciales:</b> Vincula tus hallazgos con los objetivos del proyecto.</li>
    <li><b>Relevancia de los Hallazgos:</b> Discute la importancia de tus resultados para la empresa o usuarios.</li>
    <li><b>Sugerencias y Recomendaciones:</b> Ofrece cualquier consejo basado en tus análisis.</li>
    <li><b>Reflexiones Finales:</b> Concluye con tus pensamientos finales y aprendizajes del proyecto.</li>
</ul>

Recuerda que una conclusión efectiva resume tus hallazgos y proporciona insights y reflexiones sobre su impacto y relevancia. ¡Sigue así y éxito en la etapa final!
</div>

# Comentario General del Revisor


<div class="alert alert-block alert-success">

<b>Comentario del revisor</b> <a class="tocSkip"></a>

¡Felicidades, Estela! Has completado con éxito el proyecto, mostrando un buen manejo de las herramientas necesarias para abordar un problema de clasificación. Tu enfoque, especialmente al explorar diferentes modelos, demuestra una sólida comprensión de los conceptos fundamentales. 

### Aspectos Positivos:

- Probaste múltiples modelos y ajustaste hiperparámetros de forma sistemática, lo cual es un buen paso para encontrar la mejor solución.
- Incorporaste un modelo de cordura, mostrando un enfoque riguroso para evaluar tus resultados.
- Segmentaste adecuadamente los datos en conjuntos de entrenamiento y validación, lo que asegura una evaluación justa del desempeño de los modelos.

### Recomendación para Seguir Aprendiendo:

Sería muy útil que investigues y te asegures de entender completamente la **diferencia entre clasificación y regresión**. Esto es fundamental para elegir las herramientas y métricas adecuadas en futuros proyectos. Por ejemplo, en este caso, el uso de `RandomForestRegressor` no es correcto, ya que el problema es de clasificación (predecir categorías: Smart o Ultra). Puedes explorar recursos como:

- Documentación oficial de scikit-learn: [Tipos de problemas](https://scikit-learn.org/stable/tutorial/machine_learning_map/index.html).
- Tutoriales de regresión y clasificación en plataformas como Kaggle o Coursera.

### Motivación Final:

¡Estás haciendo un excelente trabajo! Seguir afinando estos conceptos básicos te permitirá abordar problemas más complejos con confianza. Sigue así, y no dudes en pedir ayuda si te surgen dudas. 💪🚀😊

</div>